### DISCUSSION TITLE
Violation error when running FAST.Farm with supercontroller

### DISCUSSION THREAD

**Discussion by ghost** [OpenFAST/openfast#1511] [Q&A]
I'm trying to run FAST.Farm using the supercontroller, but it always gives me the same result below:

**************************************************************************************************
FAST.Farm

Copyright (C) 2023 National Renewable Energy Laboratory
Copyright (C) 2023 Envision Energy USA LTD

This program is licensed under Apache License Version 2.0 and comes with ABSOLUTELY NO WARRANTY.
See the "LICENSE" file distributed with this software for details.
**************************************************************************************************

FAST.Farm-Unversioned from 18704086dad861ab13daf804825da7c4b8d59428
Compile Info:
 - Compiler: Intel(R) Fortran Compiler 2021
 - Architecture: 64 bit
 - Precision: single
 - OpenMP: No
 - Date: Apr 02 2023
 - Time: 12:36:33
Execution Info:
 - Date: 04/02/2023
 - Time: 18:45:59-0300

 Heading of the FAST.Farm input file:
   FAST.Farm input file, using two turbines separated by 3D downstream and an offset of 30m, with
   a turbulent inflow given by one turbsim box
Running AWAE.
Running InflowWind.

   Reading a 101x35 grid (1000 m wide, 5 m to 345 m above ground) with a characteristic wind
   speed of 9.243 m/s. This full-field file was generated by TurbSim (from
   OpenFAST-v3.0.0-98-g7bde6b54) on 14-Sep-2021 at 11:37:08.

   **Processed 2000 time steps of 10-Hz full-field data (period of 200 seconds).
forrtl: severe (157): Program Exception - access violation
Image              PC                Routine            Line        Source**
*******************************************************************************************************************************

Maybe I'm not doing the right steps to run correctly.

1-Initially I compile the .sln files from the \vs-build directory.

2-After that, I update the useSC flag of the FAST.Farm.fstf file to "True" which is found in the path \openfast-3.4.1\reg_tests\r-test\glue-codes\fast-farm\TSinflow.

3-I add the SC_DLL.dll file path in SC_FileName.

4-Finally, I update YCMode to number 5 in NRELOffshrBsline5MW_Onshore_ServoDyn_WT1.dat and NRELOffshrBsline5MW_Onshore_ServoDyn_WT2.dat.

After these changes to the input files, start FAST.Farm at the prompt as follows
C:\openfast-3.4.1\openfast-3.4.1\build\bin>FAST.Farm_x64_Debug.exe C:\openfast-3.4.1\openfast-3.4.1\reg_tests\r-test\glue-codes\fast-farm \TSinflow\FAST.Farm.fstf

Am I doing something incorrect? Or are there additional steps that I'm not doing?
I'm very grateful if anyone who has managed to use the supercontroller properly can help me, because I've been doing this for a long time without success.

Note: Without the supercontroller FAST.Farm runs perfectly.

**Comment by jjonkman** [comment]
Dear @jsalesphi,

When enabling the super controller, you must also use an updated individual controller that is compatible with the super controller.  This is provided in the _DISCON_SC.f90_ / `DISCON_SC.dll` example.  Are you using that as well, distinct for each wind turbine within each OpenFAST model called by FAST.Farm?

Best regards,

**Comment by ghost** [comment]
Dear @jjonkman ,

I was not using it previously, but even after replacing with the compatible individual controller, the problem remains, showing the same access violation message.

I replaced the files:
NRELOffshrBsline5MW_Onshore_ServoDyn_WT1
NRELOffshrBsline5MW_Onshore_ServoDyn_WT2

---------------------- BLADED INTERFACE ---------------------------------------- [used only with Bladed Interface]
"C:\openfast-3.4.1\openfast-3.4.1\reg_tests\r-test\glue-codes\fast-farm\5MW_Baseline\ServoData\Discon_SC_WT1.dll"  DLL_FileName - Name/location of the dynamic library {.dll [Windows] or .so [Linux]} in the Bladed-DLL format (-) [used only with Bladed Interface]
"DISCON.IN"    DLL_InFile   - Name of input file sent to the DLL (-) [used only with Bladed Interface]
"DISCON"      DLL_ProcName - Name of procedure in DLL to be called (-) [case sensitive; used only with DLL Interface]
"default"     DLL_DT       - Communication interval for dynamic library (s) (or "default") [used only with Bladed Interface]
false         DLL_Ramp     - Whether a linear ramp should be used between DLL_DT time steps [introduces time shift when true] (flag) [used only with Bladed Interface]

---------------------- BLADED INTERFACE ---------------------------------------- [used only with Bladed Interface]
"C:\openfast-3.4.1\openfast-3.4.1\reg_tests\r-test\glue-codes\fast-farm\5MW_Baseline\ServoData\Discon_SC_WT2.dll"  DLL_FileName - Name/location of the dynamic library {.dll [Windows] or .so [Linux]} in the Bladed-DLL format (-) [used only with Bladed Interface]
"DISCON.IN"    DLL_InFile   - Name of input file sent to the DLL (-) [used only with Bladed Interface]
"DISCON"      DLL_ProcName - Name of procedure in DLL to be called (-) [case sensitive; used only with DLL Interface]
"default"     DLL_DT       - Communication interval for dynamic library (s) (or "default") [used only with Bladed Interface]
false         DLL_Ramp     - Whether a linear ramp should be used between DLL_DT time steps [introduces time shift when true] (flag) [used only with Bladed Interface]

**Reply by jjonkman** [reply]
Dear @jsalesphi,

Do you get the same access violation error when using the precompiled FAST.Farm executable and DLLs for Windows provided by NREL from: https://github.com/OpenFAST/openfast/releases?

Best regards,

**Comment by ghost** [comment]
Dear @jjonkman ,

Yes, even using the executable and DLLs from https://github.com/OpenFAST/openfast/releases the access violation message remains.

Best regards,

**Reply by jjonkman** [reply]
Dear @jsalesphi

Are you use the sample super controller (_sc_dll_) from my post dated March 18, 2021 in the following forum topic: https://forums.nrel.gov/t/fast-farm/1887 or where did you obtain the super controller?

Best regards,

**Reply by ghost** [reply]
Dear @jjonkman 

I'm using the same one you posted on the forum.

**Reply by jjonkman** [reply]
Dear @jsalesphi,

I was able to follow your process and run the example without error using the precompiled Windows binary executable of FAST.Farm (https://github.com/OpenFAST/openfast/releases/download/v3.4.1/FAST.Farm_x64.exe) and the precompiled Windows binary DLL of DISCON_SC (https://github.com/OpenFAST/openfast/releases/download/v3.4.1/Discon_SC.dll) and my own Visual Studio compile of _SC_DLL.dll_.  Perhaps it is your _SC_DLL.dll_ file that is causing the issue?  You can find my _SC_DLL.dll_ file here: https://drive.google.com/file/d/1nYYJhZzQzEa1q788Cpe_7wPUZTQfwaDj/view?usp=share_link.  Does that work for you?

Best regards,

**Reply by ghost** [reply]
Dear @jjonkman 

Even making all the replacements with the files you sent me, the violation message still appears.
I'll try to run the process on another computer and get back to you.
Thank you for your willingness to help me so far.

**Comment by j-salesps** [comment]
Dear @jjonkman,

I resumed the attempts and using your SC_DLL.dll compilation I finally got it to run correctly.

Thank you very much for your help, it will be fundamental for the progress of my research.

Best regards.

**Comment by clun1** [comment]
Dear @jjonkman,

How to use to_sc in multiple wind turbines to send parameters to the supercontroller? 
![image](https://github.com/OpenFAST/openfast/assets/93849049/5253b1e1-6786-4060-95be-e11c3d0ef8ee)
![image](https://github.com/OpenFAST/openfast/assets/93849049/e54117d7-94c2-4ead-9133-1fe753cd9043)
![image](https://github.com/OpenFAST/openfast/assets/93849049/35cfc8d4-4b59-4b70-bdbd-776b170edd0a)

could you give me some example?thank you!

Best regards.

**Reply by jjonkman** [reply]
Dear @clun1,

As mentioned above, a sample super controller for FAST.Farm is provided in my post dated March 18, 2021 in the following topic on our forum: https://forums.nrel.gov/t/fast-farm/1887.

All the super controller example does is send distinct sinusoidal yaw rates to the individual wind turbine DISCON controllers (each wind turbine has a different yaw rate).  The DISCON controllers, in turn, send the yaw rate commands to OpenFAST.  The DISCON controller also sends the generator torque of each wind turbine to the super controller, but the super controller does not make use of this data.  You should be able to execute this super controller example in any FAST.Farm simulation.

Best regards,

**Comment by 1codecopyer** [comment]
Hello,@jjonkman
I made a similar error to this colleague. The screenshot of the error is as follows:
![image](https://github.com/user-attachments/assets/1d71b74e-e3b9-45b3-99a1-2338219a3618)

[https://github.com/OpenFAST/openfast/issues/2499](url), As mentioned earlier, I was able to create a 2.2 MW stand-alone model in OpenFAST (so far, the run has not detected an error) . And I use FAST. Farm to simulate a Wind Farm consisting of four 2.2 MW WTs. I used FAST. Farm's `SC` module, which SC does not compile itself, uses NREL's official compiled super controller,[https://github.com/OpenFAST/openfast/releases/download/v3.5.3/Discon_SC.dll](url)，Use the main input file  `FAST.Farm. fstf` located in the 'reg tests \ glue codes \ fast.farm \ ModAmb_3' folder, make input parameter changes. The only change is the total simulation time `TMAX`, which has been changed to 60 seconds; And opened `UseSC` to True, and specified the file path of SC. The detailed changes are shown in the following figure:：
![image](https://github.com/user-attachments/assets/c5012df7-01f5-4b67-9d6b-c8f97a5a5d7d)
Each WT in the wind farm uses the original 2.2 MW single-unit model, and uses the corresponding sub-modules such as AD, Ed and SD module, where the SD module 'DISCON_WT.dll' file matches the 2.2 MW WT.
To check the comparison, run `FAST.Farm. fstf ` input file, keeping all other parameters unchanged, only disabling the use of SC. The result shows that FAST can run normally FAST.Farm。 I have carefully observed and browsed your response to the previous colleague, and I believe it is not a compatibility issue with the version, because `FAST.Farm-x64. exe` and `Discon_SC. dll` are both V3.5.3 versions.It doesn't make much sense to recompile `vs build \ SC_DLL \ SC_Dll.sln` using VS to generate `Discon_SC.dll` file, of course, this is just my personal guess.Because I did not significantly change the input parameters in this example, I am not sure where the problem lies. I hope you can provide some suggestions.Thanks a lot,@jjonkman

**Reply by jjonkman** [reply]
Dear @1codecopyer,

If I understand correctly, you took _DISCON_SC.dll_ from the OpenFAST v3.5.3 release and you took _SC_DLL.dll_ from my Google drive (linked above); is that correct?

My guess is the issue is that you specified the name / path to the super controller in the FAST.Farm input file (`SC_Filename`) to `DISCON_SC.dll`, rather than `SC_DLL.dll`.  To use this super controller example, `SC_Filename` in the FAST.Farm input file should point to _SC_DLL.dll_ and `DLL_FileName` in the ServoDyn input file of each wind turbine should point to _DISCON_SC.dll_.  That is, _SC_DLL.dll_ is the super controller and _DISCON_SC.dll_ is the individual turbine controller that interacts with the super controller.

That said, once you resolve this issue, please note that _SC_DLL.dll_ is only a simple example that you'll like to need to change and recompile to make use of.

Best regards,

**Comment by 1codecopyer** [comment]
Hello, @jjonkman 
Indeed, as you said, the controller for `SC_Filename` in FAST.Farm should be `SC_DLL.dll`. And the **standalone controllers** of each WT  in the wind farm need to use `Discon_SC.dll` in order to be compatible with the **wind farm level** controller `SC_Dll.dll`.
I browsed through the official help documentation of OpenFAST[ FAST.Farm SC Controller ](https://openfast.readthedocs.io/en/main/source/user/fast.farm/Introduction.html), the introduction to the super controller `SC` is relatively brief, and the content is expressed as follows: The SC module of FAST.Farm – essentially identical to the super controller available in [SOWFA](https://github.com/NREL/SOWFA) allows **wind-farm-wide control** logic to be implemented by the user, including sending and receiving commands from the individual turbine controllers in OpenFAST. The logic of such a super controller could be developed through the application of the NREL code [FLORIS](https://github.com/NREL/floris).
Therefore, I would like to inquire if you have any recommended help documents or tutorials that can help me get started with the combined use of FLORIS or SOWFA with SC, and thus assist me in conducting performance comparison research before and after actual wind farm level control.
Thanks a lot, @jjonkman

**Reply by jjonkman** [reply]
Dear @1codecopyer,

Wind farm controls are often implemented by defining a table of yaw offsets as a function of conditions (e.g., wind speed and wind direction) derived from FLORIS such that the wind farm super controller can use a table look-up to derive the optimal yaw angles as a function of conditions.

Ideally, NREL would develop a reference open-source super controller, similar to ROSCO for individual wind turbine controller, but focused on farm-level super control approaches.  However, NREL has not yet released such a reference open-source wind farm super controller.  So currently, you'll have to implement your own logic for your own purposes, using the example super controller as a template.

That said, NREL has been developing the Wind-Hybrid-Open-Controller (WHOC) that contains standard wind farm control approaches, including wake-steering, spatial-averaged consensus control, and farm level power tracking (https://github.com/nrel/wind-hybrid-open-controller).  While WHOC has not yet been interfaced to FAST.Farm, that is the plan for the future.

Best regards,

**Reply by salicto** [reply]
Dear @jjonkman,

I would like to revisit the concept of the supercontroller. To summarize briefly, I encountered issues with the error message: **Wake model is not valid in the propeller-brake region, i.e., Ct > 2.0**. 
After reading through various issues on the NREL forum and GitHub, it seems the problem stems from the supercontroller that needs to be added in Fast.Farm.fstf.

I am using Linux, and after compiling OpenFAST, I indeed obtain libsclib.so for the supercontroller module. When I specify the path and set the flag to True in the input file, I get another error message :
 **At line 303 of file /home/user/OpenFAST/ROSCO/rosco/controller/src/ReadSetParameters.f90 (unit = 13) Fortran runtime error: Cannot open file '': No such file or directory**

In my OpenFAST ServoDyn files, I have created two distinct files for turbine 1 and turbine 2. Both .so libraries were created with ROSCO. I would like to know more about this issue. 

How can I make the FAST.FARM supercontroller work with the ROSCO controller? Is this even possible? If not, how can I get the supercontroller to work without using Windows DLLs?

**************************
FAST.Farm-v4.0.4-dirty
Compile Info:

    Compiler: GCC version 10.2.1 20210110
    Architecture: 64 bit
    Precision: double
    OpenMP: Yes, number of threads: 22/22
    Date: May 19 2025
    Time: 11:39:47 Execution Info:
    Date: 05/27/2025
    Time: 14:32:32+0200
************************************
Thank you in advance for your response.

Best Regards,

**Reply by jjonkman** [reply]
Dear @salicto,

You shouldn't need a super controller in FAST.Farm to avoid errors regarding Ct > 2.  You'd have to clarify your FAST.Farm simulation set-up for me to comment on why you are receiving this error.  Are your wind turbines (OpenFAST models) operating properly up until this error is triggered?

Given that a super controller is user-defined in FAST.Farm, you've have to modify ROSCO if you want to connect ROSCO controllers for individual turbines with your own user-defined super controller.

Regarding the error message you are receiving, can you clarify what is generating this error?

Best regards,

**Reply by salicto** [reply]
Dear @jjonkman,


I will try to briefly summarize all the tests I have conducted. Initially, the same configuration worked with OpenFAST 3.5. However, after pulling the latest version 4.0.4, an error appears at t=572 s. I am not using the BeamDyn module. I initialized the rotor speed and blade pitch following the data from the IEA 22MW report. I also added wake control in the DISCON file.

I adhered to the model guidance proposed by the FAST.Farm documentation for DT_low, DT_high, DS_low, and DS_high. I used mod_ambient=2 with inflow generated by TurbSim.

I am working with the IEA 22MW model with the semi-submersible platform. After plotting some basic data, I did not find any particular behavior that indicates an error. You can find the configuration of my setup for FAST.Farm in the attached picture.

I thought that the wake dynamics might disturb the ROSCO controller and it might not manage the speed deficit properly. That is why I decided to add the super controller. What can I change in the inflow module parameters, such as reducing DT_low?
![configuration](https://github.com/user-attachments/assets/59a6076b-cf7c-47f5-b20b-9211f7a29250)
![configuration_ambientModule](https://github.com/user-attachments/assets/5434d026-91ae-40a9-a21d-655ad4257517)
![output](https://github.com/user-attachments/assets/50e07119-342d-4b68-9e63-2725513769b4)
![output_ct](https://github.com/user-attachments/assets/72f3d095-b95a-4d52-9235-4e849a90f709)

Best Regards,

**Reply by jjonkman** [reply]
Dear @salicto,

Thanks for clarifying.  You mention that you are following the model guidance, but to follow all of the FAST.Farm modeling guiance, we generally recommend using `Mod_AmbWind` = 3 together with the Python scripts available in the OpenFAST Toolbox (https://github.com/OpenFAST/openfast_toolbox/tree/main/openfast_toolbox/fastfarm) when setting up FAST.Farm simulations with TurbSim.

In your plots, can you clarify what Ct_averaged means?  Which output is this from OpenFAST or FAST.Farm?  Also, what do you have `f_c` set to in FAST.Farm?

Best regards,

**Reply by salicto** [reply]
Dear @jjonkman,

Sorry for my late response; I took some days off. I am taking a closer look at the Python files you listed. I tried to use mod_Ambient=3, but I didn't know if I needed to create separate wind files for low grid resolution and high grid resolution. I'm going to take some time to better understand this part.

Regarding the frequency, I followed Equation 4.219 from the FAST.Farm documentation with:
τ_1=1.1/(1-1.3min⁡(α_avg,0.5))  R/U,f_c=2.4/(τ_1)

The term "Ct_averaged" refers to the rotor-disk-averaged thrust coefficient for Turbine 01 (blue) and the rotor-disk-averaged thrust coefficient for Turbine 02 (orange), respectively. These two variables come from the FAST.Farm.

Best Regards,

**Reply by jjonkman** [reply]
Dear @salicto,

Thanks for clarifying regarding Ct_averaged.

I agree that `Mod_WakeAmb` = 3 requires distinct TurbSim-generated wind data files for the low- and high-resolution domains.  The creation of these distinct, but time-correlated, TurbSim-generated wind data files is automated by the Python scripts available in the OpenFAST Toolbox.

I'm guessing your cut-off frequency (`f_c`) in FAST.Farm for the wake dynamics module is set too high for the floating offshore wind turbine (FOWT) you are using.  In our work on floating wind turbine wakes with FAST.Farm: https://wes.copernicus.org/articles/9/1827/2024/wes-9-1827-2024-discussion.html, we concluded that the recommended value of `f_c` works for land-based wind turbines, but is too high for FOWT.  For FOWT, we concluded `f_c` should be reduced to, e.g., to the natural frequency of the flaoter-surge mode.

Best regards,

**Reply by salicto** [reply]
Dear @jjonkman,

You were right; changing the cutoff frequency allowed the simulation to finally converge.

I am currently experiencing some challenges with the Python toolbox for case creation, but I am confident I will overcome them. This topic was a bit off from the initial issue. I don't want to pollute this thread further with off-topic questions, so I'll make sure to start a new discussion if needed.

Thank you for your time and advice.

Best regards,
